hummingbot/market/kraken/kraken_api_order_book_data_source.py

```diff
@@ -0,0 +1,299 @@
+                            "event": "subscribe",
+                            "pair": [trading_pair],
+                            "subscription": {"name": "trade"}
+                        }
+                        await ws.send(json.dumps(subscribe_request))
+
+                    async for raw_msg in self._inner_messages(ws):
+                        msg = ujson.loads(raw_msg)
+                        if "pong" in msg:
+                            await ws.send({"event": "ping"})
+                        if type(msg) is list:
+                            if len(msg) == 4:
+                                for data in msg[1]:
+                                    trade_message: OrderBookMessage = KrakenOrderBook.trade_message_from_exchange(data, metadata={"symbol": msg[3]})
+                                    output.put_nowait(trade_message)
+                        # Server heartbeat sent if no subscription traffic within 1 second (approximately)
+                        elif msg["event"] == 'heartbeat':
+                            continue
+                        else:
+                            self.logger().debug(f"Unrecognized message received from Kraken websocket: {msg}")
+            except asyncio.CancelledError:
+                raise
+            except Exception:
+                self.logger().error("Unexpected error with WebSocket connection. Retrying after 30 seconds...",
+                                    exc_info=True)
+                await asyncio.sleep(30.0)
+
+    async def listen_for_order_book_diffs(self, ev_loop: asyncio.BaseEventLoop, output: asyncio.Queue):
+        while True:
+            try:
+                trading_pairs: List[str] = await self.get_trading_pairs()
+                async with websockets.connect(KRAKEN_WS_URI) as ws:
+                    ws: websockets.WebSocketClientProtocol = ws
+                    for trading_pair in trading_pairs:
+                        subscribe_request: Dict[str, Any] = {
+                            "event": "subscribe",
+                            "pair": [trading_pair],
+                            "subscription": {"name": "book"}
+                        }
+                        await ws.send(json.dumps(subscribe_request))
+
+                    async for raw_msg in self._inner_messages(ws):
+                        msg = ujson.loads(raw_msg)
+                        if "pong" in msg:
+                            await ws.send({"event": "ping"})
+                        if type(msg) is list:
+                            if (len(msg) == 4 or len(msg) == 5) and ('a' in msg[1] or 'b' in msg[1]):
+                                order_book_message: OrderBookMessage = KrakenOrderBook.diff_message_from_exchange(
+                                    msg,
+                                    time.time())
+                                output.put_nowait(order_book_message)
+                            elif len(msg) == 4 and 'as' in msg[1] and 'bs' in msg[1]:
```

Comment:
Since there is a way to get snapshots from Kraken WebSocket API, it would be recommended that you use the WebSocket to get the snapshots as well. Since they include timestamps for each snapshot and subsequent payload, it would be useful to use those to keep the local orderbook as close to the actual orderbook on the exchange.

You might want to look at the Bittrex connector since it does have a similar implementation.